keps/sig-node/2823-pod-addmission-plugin/README.md

```diff
@@ -0,0 +1,445 @@
+in the EKS user maintaining full host-level administrative rights on the worker
+nodes. The final option — the option that motivated this proposal — is for the
+EKS user to forego worker node management entirely using AWS Fargate, a
+serverless computing environment. With AWS Fargate, the EKS user does not have
+host-level administrative access to their worker node; in fact, the worker node
+runs on a serverless computing platform that abstracts away the entire notion of
+a host.
+
+In building the AWS EKS support for AWS Fargate, the AWS Kubernetes engineering
+team faced a dilemma: how could they prevent Pods destined to run on Fargate
+nodes from using host networking or assuming elevated host user privileges? This
+is required to restrict pods which are not supported on these worker nodes and
+also to prevent pods from accessing some portion of file system on the node or
+tampering other process on the host.
+
+The team initially investigated using a webhook or Pod Security Policy (PSP)
+that would prevent Pods with a Fargate scheduler type from having an elevated
+security context or using host networking. However, because the EKS user has
+administrative rights on the Kubernetes cluster, API-layer constructs such as a
+Pod Security Policy or validation webhook may be deleted, which would
+effectively disable the effect of that PSP. Likewise, the second solution the
+team landed on — using Node taints and tolerations — was similarly bound to the
+Kubernetes API layer, which meant EKS users could modify those Node taints and
+tolerations, effectively disabling the effects. A third potential solution
+involving OCI hooks was then investigated. OCI hooks are separate executables
+that an OCI-compatible container runtime invokes that can modify the behaviour
+of the containers in a sandbox. While this solution would have solved the
+API-layer problem, it introduced other issues, such as the inefficiency of
+downloading the container image to the Node before the OCI hook was run. Since
+pods can be scheduled directly on the node using nodename parameter in pod spec,
+its required to add such validation in Kubelet itself.
+
+The final solution the EKS team settled on involved changing kubelet itself to
+support additional node-local Pod admission checks. This KEP outlines the EKS
+team’s approach and proposes upstreaming these changes to kubelet in order to
+allow extensible node-local last-minute validation checks. This functionality
+will enable cloud providers who support nodeless/serverless worker nodes to
+restrict Pods based on fields other than those already being validated by
+kubelet.
+
+Also, Kubernetes scheduler is extensible but isn’t omniscient about everything
+on the node hence there can be cases where a scheduling decision cannot be
+fulfilled by the node based on the current node conditions such as node-level
+resource allocation. The in-tree PodAdmitHandlers mentioned below are good
```

Comment:
How about adding a new scheduler plugin that checks the POD against the host's access control attributes/policy? You can register the Fargate nodes' attributes/policy to the K8S control plane through the node feature discovery path. Doing this validation at the scheduler point is more efficient since it prevents the scheduler from selecting the Fargate nodes and eliminates all the unnecessary interaction between the K8S control plane and the kubelet of the Fargate node.

Reply:
https://github.com/kubernetes/website/issues/21128#issue-623274576 shows that scheduler plugins can be disabled.

In case of Fargate, the fleet is owned by Fargate but the customer has admin rights to the cluster. A customer can chose to disable the scheduler plugin.